src/control/server/server.go

```diff
@@ -38,167 +32,67 @@ import (
+		return nil, err
 	}
+	log.Debugf("fault domain: %s", fd.String())
 
-	iommuDisabled := !iommuDetected()
-	// Perform an automatic prepare based on the values in the config file.
-	prepReq := bdev.PrepareRequest{
-		// Default to minimum necessary for scan to work correctly.
-		HugePageCount: minHugePageCount,
-		TargetUser:    runningUser.Username,
-		PCIWhitelist:  strings.Join(cfg.BdevInclude, " "),
-		PCIBlacklist:  strings.Join(cfg.BdevExclude, " "),
-		DisableVFIO:   cfg.DisableVFIO,
-		DisableVMD:    cfg.DisableVMD || cfg.DisableVFIO || iommuDisabled,
-		// TODO: pass vmd include/white list
-	}
-
-	if cfgHasBdev(cfg) {
-		// The config value is intended to be per-engine, so we need to adjust
-		// based on the number of engines.
-		prepReq.HugePageCount = cfg.NrHugepages * len(cfg.Engines)
-
-		// Perform these checks to avoid even trying a prepare if the system
-		// isn't configured properly.
-		if runningUser.Uid != "0" {
-			if cfg.DisableVFIO {
-				return FaultVfioDisabled
-			}
-
-			if iommuDisabled {
-				return FaultIommuDisabled
-			}
-		}
-	}
+	return fd, nil
+}
 
-	log.Debugf("automatic NVMe prepare req: %+v", prepReq)
-	if _, err := bdevProvider.Prepare(prepReq); err != nil {
-		log.Errorf("automatic NVMe prepare failed (check configuration?)\n%s", err)
-	}
+// server struct contains state and components of DAOS Server.
+type server struct {
+	log         *logging.LeveledLogger
+	cfg         *config.Server
+	faultDomain *system.FaultDomain
+	ctlAddr     *net.TCPAddr
+	netDevClass uint32
+	listener    net.Listener
+
+	harness      *EngineHarness
+	membership   *system.Membership
+	sysdb        *system.Database
+	pubSub       *events.PubSub
+	evtForwarder *control.EventForwarder
+	evtLogger    *control.EventLogger
+	ctlSvc       *ControlService
+	mgmtSvc      *mgmtSvc
+	scmProvider  *scm.Provider
+	bdevProvider *bdev.Provider
+	grpcServer   *grpc.Server
+}
 
-	hugePages, err := getHugePageInfo()
+func newServer(ctx context.Context, log *logging.LeveledLogger, cfg *config.Server, fd *system.FaultDomain) (*server, error) {
+	dbReplicas, err := cfgGetReplicas(cfg, net.ResolveTCPAddr)
 	if err != nil {
-		return errors.Wrap(err, "unable to read system hugepage info")
-	}
-
-	if cfgHasBdev(cfg) {
-		// Double-check that we got the requested number of huge pages after prepare.
-		if hugePages.Free < prepReq.HugePageCount {
-			return FaultInsufficientFreeHugePages(hugePages.Free, prepReq.HugePageCount)
-		}
-	}
-
-	var dbReplicas []*net.TCPAddr
-	for _, ap := range cfg.AccessPoints {
-		apAddr, err := net.ResolveTCPAddr("tcp", ap)
-		if err != nil {
-			return config.FaultConfigBadAccessPoints
-		}
-		dbReplicas = append(dbReplicas, apAddr)
+		return nil, errors.Wrap(err, "retrieve replicas from config")
 	}
 
 	// If this daos_server instance ends up being the MS leader,
 	// this will record the DAOS system membership.
 	sysdb, err := system.NewDatabase(log, &system.DatabaseConfig{
```

Comment:
I recommend moving the sysdb/membership stuff into a method on the server struct. One problem we have with the existing codebase that is making DAOS-6004 very difficult to resolve cleanly is the fact that we can't easily reset the system database/membership. If we had a method on the server struct that configured/set the sysdb/membership fields independently of creating the server, that would make life easier in the long run I think.